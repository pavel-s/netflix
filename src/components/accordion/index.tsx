import {
  createContext,
  Dispatch,
  SetStateAction,
  useContext,
  useState,
} from 'react';
import { TRSC } from '../../types';
import {
  Container,
  Inner,
  Body,
  Frame,
  Header,
  Item,
  Title,
} from './styles/accordion';

type TAccordion = TRSC<typeof Container> & {
  Title: TRSC<typeof Title>;
  Frame: TRSC<typeof Frame>;
  Item: TRSC<typeof Item>;
  Header: TRSC<typeof Header>;
  Body: TRSC<typeof Body>;
};

const Accordion: TAccordion = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

const ToggleContext = createContext<{
  show?: boolean;
  setShow?: Dispatch<SetStateAction<boolean>>;
}>({});

Accordion.Title = ({ children, ...rest }) => (
  <Title {...rest}>{children}</Title>
);

Accordion.Frame = ({ children, ...rest }) => (
  <Frame {...rest}>{children}</Frame>
);

Accordion.Item = function AccordionItem({ children, ...rest }) {
  const [show, setShow] = useState(false);

  return (
    <ToggleContext.Provider value={{ show, setShow }}>
      <Item {...rest}>{children}</Item>
    </ToggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...rest }) {
  const { show, setShow } = useContext(ToggleContext);

  const toggleShow = () => {
    setShow && setShow((show: boolean) => !show);
  };

  return (
    <Header onClick={toggleShow} {...rest}>
      {children}{' '}
      {show ? (
        <img src='/images/icons/close-slim.png' alt='' />
      ) : (
        <img src='/images/icons/add.png' alt='' />
      )}
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...rest }) {
  const { show } = useContext(ToggleContext);
  return (
    <Body className={show ? 'open' : 'closed'}>
      <span {...rest}>{children}</span>
    </Body>
  );
};

export default Accordion;
