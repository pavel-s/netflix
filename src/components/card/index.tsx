import { TRSC } from '../../types';

import {
  Container,
  CardContainer,
  Group,
  Title,
  Items,
  Item,
  ItemImage,
  Meta,
  Subtitle,
  Text,
  Feature,
  Maturity,
  FeatureClose,
} from './styles/card';

type TCard = TRSC<typeof Container> & {
  Container: TRSC<typeof Container>;
  Group: TRSC<typeof Group>;
  Title: TRSC<typeof Title>;
  Items: TRSC<typeof Items>;
  Item: TRSC<typeof Item>;
  ItemImage: TRSC<typeof ItemImage>;
  Meta: TRSC<typeof Meta>;
  Subtitle: TRSC<typeof Subtitle>;
  Text: TRSC<typeof Text>;
  Feature: TRSC<typeof Feature>;
  Maturity: TRSC<typeof Maturity>;
  FeatureClose: TRSC<typeof FeatureClose>;
};

const Card: TCard = ({ children, ...rest }) => {
  return <CardContainer {...rest}>{children}</CardContainer>;
};

Card.Container = function CardCardContainer({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Card.Group = function CardGroup({ children, ...rest }) {
  return <Group {...rest}>{children}</Group>;
};

Card.Title = function CardTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Card.Items = function CardItems({ children, ...rest }) {
  return <Items {...rest}>{children}</Items>;
};

Card.Item = function CardItem({ children, ...rest }) {
  return <Item {...rest}>{children}</Item>;
};

Card.ItemImage = function CardItemImage({ children, ...rest }) {
  return <ItemImage {...rest}>{children}</ItemImage>;
};

Card.Meta = function CardMeta({ children, ...rest }) {
  return <Meta {...rest}>{children}</Meta>;
};

Card.Subtitle = function CardSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

Card.Text = function CardText({ children, ...rest }) {
  return <Text {...rest}>{children}</Text>;
};

Card.Feature = function CardFeature({ children, ...rest }) {
  return <Feature {...rest}>{children}</Feature>;
};

Card.Maturity = function CardMaturity({ maturity, ...rest }) {
  return (
    <Maturity maturity={maturity} {...rest}>
      {maturity}
    </Maturity>
  );
};

Card.FeatureClose = function CardFeatureClose({ children, ...rest }) {
  return <FeatureClose {...rest}>{children}</FeatureClose>;
};

export default Card;
