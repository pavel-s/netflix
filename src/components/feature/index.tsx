import { TRSC } from '../../types';
import { Container, Inner, Title, Subtitle } from './styles/feature';

type TFeature = TRSC<'div'> & {
  Title: TRSC<'h1'>;
  Subtitle: TRSC<'h2'>;
};

const Feature: TFeature = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

Feature.Title = function FeatureTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Feature.Subtitle = function FeatureSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

export default Feature;
