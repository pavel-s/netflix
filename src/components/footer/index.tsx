import { TRSC } from '../../types';
import { Container, Inner, Title, Links, Link, Text } from './styles/footer';

type TFooter = TRSC<'div'> & {
  Title: TRSC<'p'>;
  Links: TRSC<'ul'>;
  Link: TRSC<'a'>;
  Text: TRSC<'p'>;
};

const Footer: TFooter = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

Footer.Title = ({ children, ...rest }) => {
  return <Title {...rest}>{children}</Title>;
};

Footer.Links = ({ children, ...rest }) => {
  return <Links {...rest}>{children}</Links>;
};

Footer.Link = ({ children, ...rest }) => {
  return (
    <Link>
      <a {...rest}>{children}</a>
    </Link>
  );
};

Footer.Text = ({ children, ...rest }) => {
  return <Text {...rest}>{children}</Text>;
};

export default Footer;
