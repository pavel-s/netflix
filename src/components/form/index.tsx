import { TRSC } from '../../types';
import {
  Container,
  Base,
  Title,
  Input,
  Error,
  Submit,
  Row,
  Checkbox,
  Text,
  TextBig,
  Link,
  TextButton,
} from './styles/form';

type TForm = TRSC<typeof Container> & {
  Base: TRSC<typeof Base>;
  Title: TRSC<typeof Title>;
  Input: TRSC<typeof Input>;
  Error: TRSC<typeof Error>;
  Submit: TRSC<typeof Submit>;
  Row: TRSC<typeof Row>;
  Checkbox: TRSC<typeof Checkbox>;
  Text: TRSC<typeof Text>;
  TextBig: TRSC<typeof TextBig>;
  Link: TRSC<typeof Link>;
  TextButton: TRSC<typeof TextButton>;
};

const Form: TForm = ({ children, ...rest }) => {
  return <Container {...rest}>{children}</Container>;
};

Form.Base = function FormBase({ children, ...rest }) {
  return <Base {...rest}>{children}</Base>;
};

Form.Title = function FormTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Form.Input = function FormInput({ ...rest }) {
  return <Input {...rest} />;
};

Form.Error = function FormError({ ...rest }) {
  return <Error {...rest} />;
};

Form.Submit = function FormSubmit({ isLoading = false, children, ...rest }) {
  return (
    <Submit {...rest} isLoading={isLoading} type='submit'>
      {children}
    </Submit>
  );
};

Form.Row = function FormRow({ children, ...rest }) {
  return <Row {...rest}>{children}</Row>;
};

Form.Checkbox = function FormCheckbox({ children, ...rest }) {
  return <Checkbox {...rest}>{children}</Checkbox>;
};

Form.Text = function FormText({ children, ...rest }) {
  return <Text {...rest}>{children}</Text>;
};

Form.TextBig = function FormTextBig({ children, ...rest }) {
  return <TextBig {...rest}>{children}</TextBig>;
};

Form.Link = function FormLink({ children, ...rest }) {
  return <Link {...rest}>{children}</Link>;
};

Form.TextButton = function FormTextButton({ children, ...rest }) {
  return <TextButton {...rest}>{children}</TextButton>;
};

export default Form;
