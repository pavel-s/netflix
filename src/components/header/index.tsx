import { Background, Container, Logo, SignButton } from './styles/header';
import { LinkProps } from 'react-router-dom';
import { TRSC } from '../../types';

type THeader = TRSC<'div'> & {
  // Container: TRSC<'div'>;
  Logo: TRSC<'img'>;
  Frame: TRSC<'div'>;
  SignButton: TRSC<'a', LinkProps>;
};

const Header: THeader = ({ children, ...rest }) => {
  return <Background {...rest}>{children}</Background>;
};

Header.Frame = function HeaderFrame({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Header.Logo = function HeaderLogo({ children, ...rest }) {
  return <Logo {...rest} />;
};

Header.SignButton = function HeaderSignButton({ children, ...rest }) {
  return <SignButton {...rest}>{children}</SignButton>;
};

export default Header;
