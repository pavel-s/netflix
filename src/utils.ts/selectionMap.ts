import { TMediaItem } from '../types';

const genres = [
  { title: 'Children', genre: 'children' },
  { title: 'Drama', genre: 'drama' },
  { title: 'Romance', genre: 'romance' },
  { title: 'Suspense', genre: 'suspense' },
  { title: 'Thriller', genre: 'thriller' },
];

/**
 * map films or series on genres
 */
export const selectionMap = (items: TMediaItem[]) => {
  const result: { [genre: string]: { title: string; items: TMediaItem[] } } =
    {};

  genres.forEach(({ genre, title }) => {
    result[genre] = {
      title,
      items: items.filter((item) => item.genre === genre),
    };
  });

  return result;
};
