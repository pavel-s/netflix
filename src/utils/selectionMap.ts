import { TMediaItem } from '../types';

const genres = [
  { title: 'Children', genre: 'children' },
  { title: 'Drama', genre: 'drama' },
  { title: 'Romance', genre: 'romance' },
  { title: 'Suspense', genre: 'suspense' },
  { title: 'Thriller', genre: 'thriller' },
  { title: 'Comedies', genre: 'comedies' },
  { title: 'Crime', genre: 'crime' },
  { title: 'Documentaries', genre: 'documentaries' },
  { title: 'Feel good', genre: 'feel-good' },
];

type TSelectionResultItem = { title: string; items: TMediaItem[] };

/**
 * map films or series on genres
 */
export const selectionMap = (items: TMediaItem[]) => {
  const result: TSelectionResultItem[] = [];

  genres.forEach(({ genre, title }) => {
    result.push({
      title,
      items: items.filter((item) => item.genre === genre),
    });
  });

  return result.filter((genre) => genre.items.length > 0);
};
