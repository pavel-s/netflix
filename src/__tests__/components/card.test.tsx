import { render } from '@testing-library/react';
import { Card } from '../../components';
import { selectionMap } from '../../utils.ts/selectionMap';

// data
const rows: ReturnType<typeof selectionMap> = [
  {
    title: 'Children',
    items: [
      {
        title: 'Joker',
        description: 'Joker description',
        docId: 'sdfsdfsdf33f3ff3f',
        id: 'sdfsdfsdf3f3f3f',
        maturity: '15',
        slug: 'joker',
        genre: 'children',
      },
    ],
  },
];

const category = 'films';
const row = rows[0];
const item = rows[0].items[0];

describe('<Card />', () => {
  it('render Card with populated data', () => {
    const { container, queryByText, queryByAltText } = render(
      <Card.Container>
        <Card>
          <Card.Title>{row.title}</Card.Title>
          <Card.Items>
            <Card.Item>
              <Card.ItemImage
                src={`/images/${category}/${item.genre}/${item.slug}/small.jpg`}
                alt={item.title}
              />
              <Card.Meta>
                <Card.Subtitle>{item.title}</Card.Subtitle>
                <Card.Text>{item.description}</Card.Text>
              </Card.Meta>
            </Card.Item>
          </Card.Items>
        </Card>
      </Card.Container>
    );

    // card title
    expect(queryByText(row.title)).toBeTruthy();

    // item image
    expect(queryByAltText(item.title)).toHaveAttribute(
      'src',
      `/images/${category}/${item.genre}/${item.slug}/small.jpg`
    );

    // item title
    expect(queryByText(item.title, { selector: 'h3' })).toBeTruthy();

    // item description
    expect(queryByText(item.description)).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('render Card.Feature with populated data', () => {
    const { container, queryByText, queryByAltText } = render(
      <Card.Container>
        <Card>
          <Card.Feature
            url={`/images/${category}/${item.genre}/${item.slug}/large.jpg`}
          >
            <Card.Title>{item.title}</Card.Title>
            <Card.Text>{item.description}</Card.Text>

            <Card.Group>
              <Card.Maturity maturity={+item.maturity} />
              <Card.Subtitle>
                {item.genre[0].toLocaleUpperCase() + item.genre.slice(1)}
              </Card.Subtitle>
            </Card.Group>

            <Card.FeatureClose>
              <img src='/images/icons/close.png' alt='close' />
            </Card.FeatureClose>
          </Card.Feature>
        </Card>
      </Card.Container>
    );

    // feature title
    expect(queryByText(item.title)).toBeTruthy();

    // feature description
    expect(queryByText(item.description)).toBeTruthy();

    // feature maturity
    expect(queryByText(item.maturity)).toBeTruthy();

    // feature genre
    expect(
      queryByText(item.genre[0].toLocaleUpperCase() + item.genre.slice(1))
    ).toBeTruthy();

    // feature close button
    expect(queryByAltText('close')).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('render Card.Feature with no background and 0 maturity', () => {
    const { container, queryByText } = render(
      <Card.Container>
        <Card>
          <Card.Feature url={''}>
            <Card.Group>
              <Card.Maturity maturity={0} />
            </Card.Group>
          </Card.Feature>
        </Card>
      </Card.Container>
    );

    // feature maturity
    expect(queryByText(0)).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
