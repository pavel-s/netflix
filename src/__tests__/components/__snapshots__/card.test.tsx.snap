// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> render Card with populated data 1`] = `
<div
  class="card__Container-vait2u-0 gDjoKe"
>
  <div
    class="card__CardContainer-vait2u-1 csYGmB"
  >
    <h2
      class="card__Title-vait2u-2 fmrCPZ"
    >
      Children
    </h2>
    <div
      class="card__Items-vait2u-3 grIBiZ"
    >
      <button
        class="card__Item-vait2u-6 iMYRJV"
      >
        <img
          alt="Joker"
          class="card__ItemImage-vait2u-4 ehhpt"
          src="/images/films/children/joker/small.jpg"
        />
        <div
          class="card__Meta-vait2u-5 hzdAah"
        >
          <h3
            class="card__Subtitle-vait2u-7 ejogZR"
          >
            Joker
          </h3>
          <p
            class="card__Text-vait2u-8 hUQirf"
          >
            Joker description
          </p>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Card /> render Card.Feature with no background and 0 maturity 1`] = `
<div
  class="card__Container-vait2u-0 gDjoKe"
>
  <div
    class="card__CardContainer-vait2u-1 csYGmB"
  >
    <div
      class="card__Feature-vait2u-9 bOidRF"
    >
      <div
        class="card__Group-vait2u-10 dykZoC"
      >
        <div
          class="card__Maturity-vait2u-11 kYVsAL"
        >
          0
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Card /> render Card.Feature with populated data 1`] = `
<div
  class="card__Container-vait2u-0 gDjoKe"
>
  <div
    class="card__CardContainer-vait2u-1 csYGmB"
  >
    <div
      class="card__Feature-vait2u-9 jdhkkb"
    >
      <h2
        class="card__Title-vait2u-2 fmrCPZ"
      >
        Joker
      </h2>
      <p
        class="card__Text-vait2u-8 hUQirf"
      >
        Joker description
      </p>
      <div
        class="card__Group-vait2u-10 dykZoC"
      >
        <div
          class="card__Maturity-vait2u-11 iSkcjr"
        >
          15
        </div>
        <h3
          class="card__Subtitle-vait2u-7 ejogZR"
        >
          Children
        </h3>
      </div>
      <button
        class="card__FeatureClose-vait2u-12 tCXhS"
      >
        <img
          alt="close"
          src="/images/icons/close.png"
        />
      </button>
    </div>
  </div>
</div>
`;
